### PR TITLE
Hindi Language Numeral Dimension(minimalistic model). Tests passed.

### PR DESCRIPTION
**Pull request by NewInfinite007**

### CODE REVIEW

File: Duckling/Numeral/HI/Corpus.hs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+  , examples (NumeralValue 8)
+             [ "आठ"
+             ]
+  , examples (NumeralValue 9)
+             [ "नौ"
+             ]
+  , examples (NumeralValue 10)
+             [ "दस"
+             ]
+--  , examples (NumeralValue 15)
```
Comment (patapizza) [inline Duckling/Numeral/HI/Corpus.hs]: let's fix these examples and uncomment

File: Duckling/Numeral/HI/Rules.hs [hunk 2]
```diff
@@ -0,0 +1,181 @@
+-- Copyright (c) 2016-present, Facebook, Inc.
+-- All rights reserved.
+--
+-- This source code is licensed under the BSD-style license found in the
+-- LICENSE file in the root directory of this source tree. An additional grant
+-- of patent rights can be found in the PATENTS file in the same directory.
+
+-------------------------WORK IN PROGRESS------------------------
```
Comment (patapizza) [inline Duckling/Numeral/HI/Rules.hs]: rm

File: Duckling/Numeral/HI/Rules.hs [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+       _) -> do
+         v <- parseInt match
+         integer $ toInteger v
+      _ -> Nothing
+  }
+
+ruleNumeralMap :: HashMap Text Integer
+ruleNumeralMap = HashMap.fromList
+  [ ( "शून्य", 0)
+--  , ( "z\x00E9r\x00F3", 0 )
```
Comment (patapizza) [inline Duckling/Numeral/HI/Rules.hs]: use unicode chars directly

File: Duckling/Numeral/HI/Rules.hs [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+  , ("नौ", 9)
+--  , ( "t\x00EDz", 10)
+  , ("दस", 10)
+  ]
+
+ruleNumeral :: Rule
+ruleNumeral = Rule
+  { name = "number (0..10)"
+  , pattern =
+    [ regex "(शून्य|एक|दो|तीन|चार|पाँच|छह|सात|आठ|नौ|दस)"
```
Comment (patapizza) [inline Duckling/Numeral/HI/Rules.hs]: seems like this regex is missing western strings that are present in `ruleNumeralMap`

File: Duckling/Numeral/HI/Rules.hs [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+  , prod = \tokens -> case tokens of
+      (Token RegexMatch (GroupMatch (match:_)):_) ->
+        HashMap.lookup (Text.toLower match) ruleNumeralMap >>= integer
+      _ -> Nothing
+  }
+
+elevenToNineteenMap :: HashMap Text Integer
+elevenToNineteenMap = HashMap.fromList
+  [ ( "tizenegy", 11 )
+  , ( "tizenkett\x0151", 12 )
```
Comment (patapizza) [inline Duckling/Numeral/HI/Rules.hs]: use unicode directly

File: Duckling/Numeral/HI/Rules.hs [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+  , ( "huszonh\x00E9t", 27 )
+  , ( "huszonnyolc", 28 )
+  , ( "huszonkilenc", 29 )
+  ]
+
+ruleTwentyoneToTwentynine :: Rule
+ruleTwentyoneToTwentynine = Rule
+  { name = "number (21..29)"
+  , pattern =
+    [ regex "(huszonegy|huszonkett\x0151|huszonh\x00E1rom|huszonn\x00E9gy|huszon\x00F6t|huszonhat|huszonh\x00E9t|huszonnyolc|huszonkilenc)"
```
Comment (patapizza) [inline Duckling/Numeral/HI/Rules.hs]: unicode

Comment (patapizza) [reply Duckling/Numeral/HI/Rules.hs]: looks like this regex can be simplified into `huszon(egy|kett..)`

File: Duckling/Numeral/HI/Rules.hs [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+    ]
+  , prod = \tokens -> case tokens of
+      (Token RegexMatch (GroupMatch (match:_)):_) ->
+        HashMap.lookup (Text.toLower match) twentyoneToTwentynineMap >>= integer
+      _ -> Nothing
+  }
+
+tensMap :: HashMap Text Integer
+tensMap = HashMap.fromList
+  [ ( "h\x00FAsz", 20 )
```
Comment (patapizza) [inline Duckling/Numeral/HI/Rules.hs]: unicode

File: Duckling/Numeral/HI/Rules.hs [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+  , ( "hetven", 70 )
+  , ( "nyolcvan", 80 )
+  , ( "kilencven", 90 )
+  ]
+
+ruleTens :: Rule
+ruleTens = Rule
+  { name = "integer (20,30..90)"
+  , pattern =
+    [ regex "(h\x00FAsz|harminc|negyven|ötven|hatvan|hetven|nyolcvan|kilencven)"
```
Comment (patapizza) [inline Duckling/Numeral/HI/Rules.hs]: unicode

File: Duckling/Numeral/HI/Rules.hs [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+      (Token RegexMatch (GroupMatch (match:_)):_) ->
+        HashMap.lookup (Text.toLower match) tensMap >>= integer
+      _ -> Nothing
+  }
+
+ruleCompositeTens :: Rule
+ruleCompositeTens = Rule
+  { name = "integer ([3-9][1-9])"
+  , pattern =
+    [ regex "(harminc|negyven|\x00F6tven|hatvan|hetven|nyolcvan|kilencven)(egy|kett\x0151|h\x00E1rom|n\x00E9gy|\x00F6t|hat|h\x00E9t|nyolc|kilenc)"
```
Comment (patapizza) [inline Duckling/Numeral/HI/Rules.hs]: unicode

File: Duckling/Numeral/HI/Rules.hs [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+        v2 <- HashMap.lookup (Text.toLower m2) ruleNumeralMap
+        integer $ v1 + v2
+      _ -> Nothing
+  }
+
+rules :: [Rule]
+rules =
+  [   ruleIntegerNumeric
+    , ruleNumeral
+--  , ruleElevenToNineteen
```
Comment (patapizza) [inline Duckling/Numeral/HI/Rules.hs]: uncomment after fix

File: Duckling/Rules/HI.hs [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+-- of patent rights can be found in the PATENTS file in the same directory.
+
+-------------------------WORK IN PROGRESS------------------------
+
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Duckling.Rules.HI
+  ( defaultRules,
+    langRules
```
Comment (patapizza) [inline Duckling/Rules/HI.hs]: move comma here, align with (

File: tests/Duckling/Numeral/HI/Tests.hs [hunk 12]
```diff
@@ -0,0 +1,22 @@
+--Copyright (c) 2016-present, Facebook, Inc.
+-- All rights reserved.
+--
+-- This source code is licensed under the BSD-style license found in the
+-- LICENSE file in the root directory of this source tree. An additional grant
+-- of patent rights can be found in the PATENTS file in the same directory.
+
+module Duckling.Numeral.HI.Tests
+  ( tests ) where
```
Comment (patapizza) [inline tests/Duckling/Numeral/HI/Tests.hs]: You need to include these tests in `tests/Duckling/Numeral/Tests.hs`

**Comment by NewInfinite007** [comment]
Hi @patapizza ,

  After modifying the Corpus.hs and Rules.hs , building and running the model, I try examples like : `शून्य, एक` but I get an empty list in return. Could you please give me an insight into what exactly is happening in this case? (p.s. I am not really a Haskell programmer)

Thanks in advance :)

**Comment by NewInfinite007** [comment]
Hi @patapizza  ,

I apologise for an untidy commit. Will not happen next time.

Regards
@NewInfinite007

**Comment by patapizza** [comment]
@NewInfinite007 No worries at all!

How are you trying the examples? I see that `शून्य` is present in a `Corpus.hs` file that is part of a test, and the tests passed - so it should work.

File: .idea/duckling.iml [hunk 13]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (patapizza) [inline .idea/duckling.iml]: revert file

File: .idea/misc.xml [hunk 14]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (patapizza) [inline .idea/misc.xml]: ditto

File: .idea/modules.xml [hunk 15]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (patapizza) [inline .idea/modules.xml]: ditto

File: .idea/vcs.xml [hunk 16]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (patapizza) [inline .idea/vcs.xml]: ditto

File: Duckling/Dimensions.hs [hunk 17]
```diff
@@ -34,6 +34,7 @@ import qualified Duckling.Dimensions.GA as GADimensions
 import qualified Duckling.Dimensions.HE as HEDimensions
 import qualified Duckling.Dimensions.HR as HRDimensions
 import qualified Duckling.Dimensions.HU as HUDimensions
+import qualified Duckling.Dimensions.HI as HIDimensions
```
Comment (patapizza) [inline Duckling/Dimensions.hs]: `HI` should come between `HE` and `HR`

File: Duckling/Dimensions.hs [hunk 18]
```diff
@@ -95,6 +96,7 @@ langDimensions GA = GADimensions.allDimensions
 langDimensions HE = HEDimensions.allDimensions
 langDimensions HR = HRDimensions.allDimensions
 langDimensions HU = HUDimensions.allDimensions
+langDimensions HI = HIDimensions.allDimensions
```
Comment (patapizza) [inline Duckling/Dimensions.hs]: ditto

File: Duckling/Locale.hs [hunk 19]
```diff
@@ -45,6 +45,7 @@ data Lang
   | HE
   | HR
   | HU
+  | HI
```
Comment (patapizza) [inline Duckling/Locale.hs]: ditto

File: Duckling/Numeral/HI/Corpus.hs [hunk 20]
```diff
@@ -61,40 +57,28 @@ allExamples = concat
   , examples (NumeralValue 10)
              [ "दस"
              ]
---  , examples (NumeralValue 15)
```
Comment (patapizza) [inline Duckling/Numeral/HI/Corpus.hs]: let's get these examples working if they are valid and add them under the existing ones below

File: Duckling/Numeral/HI/Corpus.hs [hunk 21]
```diff
@@ -61,40 +57,28 @@ allExamples = concat
...
+             [ "छब्बीस"
+             ]
+  , examples (NumeralValue 28)
+             [ "अट्ठाईस"
+             ]
+  , examples (NumeralValue 50)
+             [ "पचास"
+             ]
   ]
------------------------------------------------------------------
```
Comment (patapizza) [inline Duckling/Numeral/HI/Corpus.hs]: revert

File: Duckling/Numeral/HI/Rules.hs [hunk 22]
```diff
@@ -46,34 +46,35 @@ ruleIntegerNumeric = Rule
...
+  , ( "०", 0 )
   , ( "एक", 1 )
---  , ( "kett\x0151",  2 )
-  , ( "दो", 2)
---  , ( "h\x00E1rom", 3 )
+  , ( "१", 1 )
+  , ( "दो", 2 )
+  , ( "२", 2 )
   , ( "तीन", 3 )
---  , ( "n\x00E9gy", 4 )
```
Comment (patapizza) [inline Duckling/Numeral/HI/Rules.hs]: if this is legit, add back in unicode

File: Duckling/Numeral/HI/Rules.hs [hunk 23]
```diff
@@ -83,22 +84,31 @@ ruleNumeral = Rule
 
 elevenToNineteenMap :: HashMap Text Integer
 elevenToNineteenMap = HashMap.fromList
-  [ ( "tizenegy", 11 )
```
Comment (patapizza) [inline Duckling/Numeral/HI/Rules.hs]: ditto

File: Duckling/Numeral/HI/Rules.hs [hunk 24]
```diff
@@ -108,22 +118,31 @@ ruleElevenToNineteen = Rule
 
 twentyoneToTwentynineMap :: HashMap Text Integer
 twentyoneToTwentynineMap = HashMap.fromList
-  [ ( "huszonegy", 21 )
```
Comment (patapizza) [inline Duckling/Numeral/HI/Rules.hs]: ditto

File: Duckling/Numeral/HI/Rules.hs [hunk 25]
```diff
@@ -133,49 +152,42 @@ ruleTwentyoneToTwentynine = Rule
 
 tensMap :: HashMap Text Integer
 tensMap = HashMap.fromList
-  [ ( "h\x00FAsz", 20 )
```
Comment (patapizza) [inline Duckling/Numeral/HI/Rules.hs]: ditto

File: Duckling/Ranking/Classifiers.hs [hunk 26]
```diff
@@ -30,6 +30,7 @@ import qualified Duckling.Ranking.Classifiers.GA_XX as GA_XXClassifiers
 import qualified Duckling.Ranking.Classifiers.HE_XX as HE_XXClassifiers
 import qualified Duckling.Ranking.Classifiers.HR_XX as HR_XXClassifiers
 import qualified Duckling.Ranking.Classifiers.HU_XX as HU_XXClassifiers
+import qualified Duckling.Ranking.Classifiers.HI_XX as HI_XXClassifiers
```
Comment (patapizza) [inline Duckling/Ranking/Classifiers.hs]: ditto re alphabetical order

File: duckling.cabal [hunk 27]
```diff
@@ -54,6 +54,7 @@ library
                      , Duckling.Rules.HE
                      , Duckling.Rules.HR
                      , Duckling.Rules.HU
+                     , Duckling.Rules.HI
```
Comment (patapizza) [inline duckling.cabal]: ditto

**Comment by NewInfinite007** [comment]
Hi @patapizza ,

 I got the model working. It seems I was sending wrong POST request to the server (My bad :p). There are more modifications to be made. Would it be ok if I made contribution to the repository from my personal git account? (This is my professional account. My personal git account is [infinite007)](https://github.com/infinite007)

**Comment by patapizza** [comment]
@NewInfinite007 Sure, feel free to close this one and do a new PR then. Otherwise let me know.

**Comment by patapizza** [comment]
Thanks @NewInfinite007!

**Pull request closed without merging**